### ISSUE TITLE
Che Plugin Marketplace - Phase 1

### ISSUE THREAD

**Issue by slemeur** [eclipse-che/che#8282]
### Description
A user who built his plugin should be able to distribute it to the Eclipse Che community. In order to handle the different aspects of the distribution of Eclipse Che plugins, we would be building a Plugin Marketplace.

A Plugin author should be able to publish his plugin by filling a request to the marketplace. Marketplace's administrators will then validate the request to get the plugin available to all Che users.

### Experiences
In order to publish his plugin, the plugin author will have to go to the plugin marketplace and fill a request. The request will then be sent to the plugin marketplace administrator which will review the plugin proposal, validate it or work along with the plugin’s author to get the plugin's information and request completed.

Once validated and published, Eclipse Che user will be able to see the plugin listed under workspace's settings.

### Proposed Phase 1
Until the marketplace is completed, we should be building a list of Che plugins under a public repository. This list could be based on a file with the metadatas needed for the user dashboard.

This way, publishing a plugin to Eclipse Che Marketplace would be filling a PR with the needed metadatas.

### Sub tasks:
- [x] Investigate Mozilla Addons Server repository #9863 
- [ ] Plugin Marketplace UI - Phase 1 https://github.com/eclipse/che/issues/9866
- [ ] Study Mozilla Firefox Marketplace frontend #9867 
- [x] Initial Plugin registry #9868

**Comment by vparfonov** [comment]
What about using `npm` for plugin registry?

**Comment by ghost** [comment]
It does look like npm registry has good answers for all user and contributor stories that @slemeur has mentioned. Moreover, in certain scenarios it is possible to run a local npm registry in a container and package all available plugins in there, and this container/pod will be part of Che infra. A local registyr can be also used in dev scenarios, when publishing to a global registry is premature but some testing still needs to be done.

**Comment by skabashnyuk** [comment]
Last Thursday we have a meeting
with @gorkem @slemeur @ashumilova @l0rd @vparfonov @benoitf 
We talked about a possible format for Che plugins.
At this moment we thought about zip file with a couple of resources inside.

Packaging of the Che Feature

.cfa (che feature archive)
- marketing.png
- plugin.yaml (metadata)
- feature.yaml (features)
- service.yaml(services)
- theia1Plugin.zip
- theia2Plugin.zip


proposed content of plugin.yaml
```yaml
version: "0.0.1"
name: "che-theia-ssh"
title: "SSH Client"
homepage: "https://github.com/eclipse/che-theia-ssh"
description: "Eclipse Che Plugin for SSH"
created_at: "2015-11-22T05:40:57Z"
updated_at: "2018-04-22T00:27:28Z"
installation_count: 88
license: 
 key: mit
 name: "MIT License"
 url: "https://api.github.com/licenses/mit"
owner: 
 name: redhat
 avatar_url: ""
 gravatar_id: ""
 url: "https://redhat.com"
```yaml



Our Priorities:
- List of all the plugins that we can install in a json file
- With a cfa on github (for preliminary/mocking step)

**Comment by skabashnyuk** [comment]
Proposed API for plugins
![2018-05-03 11 09 31](https://user-images.githubusercontent.com/1614429/39566789-68d24400-4ec5-11e8-8154-62e859c34bf9.png)


```POST```  method of plugin creation omitted intentionally. We will go back to it soon.
See more 
https://app.swaggerhub.com/apis/skabashniuk/Che/1.0.0#/definitions/ChePlugin

**Comment by skabashnyuk** [comment]
@benoitf @gorkem  @vparfonov here is instruction how you can test marketplace prototype
https://github.com/skabashnyuk/kubsrv/blob/master/README.md

**Comment by benoitf** [comment]
how do I get a mock plugin that includes a theia1Plugin.zip ?

**Comment by benoitf** [comment]
I will be able to provide to you soon a valid link to a theia zip plugin (so we could have for example to test
```
- theia
   - http://path-to-a-zip.file
```
inside the plugin.yaml

**Comment by skabashnyuk** [comment]
@benoitf 
As far as I understand you need to somehow grab the content of plugin to be able to start it with Theia
For example che-theia-github. This feature has such description.
```yaml
apiVersion: v1
kind: CheFeature
metadata:
  name: che-theia-github
spec:
  version: 0.0.1
  description: Theia's Github plugin
  services:
    - name : org.eclipse.che.theia-ide
      version : 0.0.1
      parameters:
          - name: THEIA_PLUGINS
            value: che-theia-github.tar.gz
```
What if I've change it on this form

```yaml
apiVersion: v1
kind: CheFeature
metadata:
  name: che-theia-github
spec:
  version: 0.0.1
  description: Theia's Github plugin
  services:
    - name : org.eclipse.che.theia-ide
      version : 0.0.1
      parameters:
          - name: THEIA_PLUGINS
            value: file:http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github.tar.gz
```            
In this case you will have url of this plugin as part of environment variable inside of theia container
Would it work for you?

**Comment by skabashnyuk** [comment]
BTW @benoitf have you tried this https://github.com/skabashnyuk/kubsrv/blob/master/README.md#how-to-use or you exeprement with swaggerhub?

**Comment by benoitf** [comment]
> @benoitf
 As far as I understand you need to somehow grab the content of plugin to be able to start it with Theia
 For example che-theia-github. This feature has such description.

I start from the description of a `.car` file that can embed theia plugins. So I want to grab this zip that was part of the `.car` file.
```
.cfa (che feature archive)

marketing.png
plugin.yaml (metadata)
feature.yaml (features)
service.yaml(services)
theia1Plugin.zip
theia2Plugin.zip
```

I need to grab these theia zip files from 'market place'

> BTW @benoitf have you tried this > https://github.com/skabashnyuk/kubsrv/blob/master/README.md#how-to-use 

I'm experimenting with swaggerHub. BTW we said that we should try with a JSON first approach.

**Comment by skabashnyuk** [comment]
My thought that main purpose for .car file is only to submit plugins(we can postpone this method and just directly add contnet to git like I did here https://github.com/skabashnyuk/che-registry/tree/master/org/eclipse/che/che-theia-github/0.0.1). I didn't expect any other methods that will return .car file.

>BTW we said that we should try with a JSON first approach.

I can convert it to json. But I guess for users when they will work on ```.car``` file content it would be easier to work with yaml. Anyway json/yaml is not a problem. I can convert everything to json. 


>I start from the description of a .car file that can embed theia plugins

I think if you want to work with plugin metadata you shuld take a look here https://github.com/skabashnyuk/che-registry/blob/master/org/eclipse/che/che-theia-github/0.0.1/CheMeta.yaml or here https://github.com/skabashnyuk/che-registry/blob/master/org/eclipse/che/che-theia-ssh/0.0.1/CheMeta.yaml

**Comment by benoitf** [comment]
> My thought that main purpose for .car file is only to submit plugins(we can postpone this method and just directly add contnet to git like I did here https://github.com/skabashnyuk/che-registry/tree/master/org/eclipse/che/che-theia-github/0.0.1). I didn't expect any other methods that will return .car file.

I'm fine to submit only '.car files" but almost all data of .car files should be available to clients.
so if I upload a .car file embedding a theia plugin, I should have a way to list these files and grab them as well, correct ?


- about metadata :
> What if I've change it on this form
 ```
apiVersion: v1
kind: CheFeature
metadata:
  name: che-theia-github
spec:
  version: 0.0.1
  description: Theia's Github plugin
  services:
    - name : org.eclipse.che.theia-ide
      version : 0.0.1
      parameters:
          - name: THEIA_PLUGINS
            value: url:http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github.tar.gz
```

in the first sample the .car contained two theia plugin zip files so value would have two URLs ? in THEIA_PLUGINS ? or it provides an array

also in my .car I didn't specify org.eclipse.che.theia-ide, it was only containing the theia zip files. it's the service that has converted and added that automatically ?

is `/plugin/version/`

**Comment by skabashnyuk** [comment]
>I'm fine to submit only '.car files" but almost all data of .car files should be available to clients.

In my meaning it is. 
You can get 
1. plugin metadata with ```curl  http://localhost:3000/plugin/org.eclipse.che.che-theia-ssh/0.0.1```
2. plugin feature with ```curl  http://localhost:3000/feature/org.eclipse.che.che-theia-ssh/0.0.1```
3. plugin service with ```curl  http://localhost:3000/service/org.eclipse.che.theia-ide/0.0.1```


>so if I upload a .car file embedding a theia plugin, I should have a way to list these files and grab them as well, correct ?
>

About everything else let's talk on concrete examples, what you want to get and why?


>in the first sample the .car contained two theia plugin zip files so value would have two URLs ? in THEIA_PLUGINS ? or it provides an array

I thought about something like this.

```
     parameters:
         - name: THEIA_PLUGINS
           value: url:http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github.tar.gz,url:http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github2.tar.gz
```           
But we can imagine any other logical format if you want

>also in my .car I didn't specify org.eclipse.che.theia-ide, it was only containing the theia zip files. it's the service that has converted and added that automatically ?

AFAIK we talked about some cli tool.

**Comment by benoitf** [comment]
> About everything else let's talk on concrete examples, what you want to get and why?

I'm talking about getting the two theia zip plugins of the .car since the beginning

>I thought about something like this.
```
     parameters:
         - name: THEIA_PLUGINS
           value: url:http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github.tar.gz,url:http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github2.tar.gz
```
My question is why is it in a so hidden output format for clients that need to grab the two theia plugin zip files ?

a client would need to ask the list of services then look if there is service named org.eclipse.che.theia-ide
then, grab THEIA_PLUGINS and then parse again a comma separated value
not so easy at all.


Also :
1.how can I see the description of these two theia plugins ? you're only exposing a link to the URL but these zip files contain some description and metadata as well like the engine required to run, etc.
so In order to know if this plugin is possible to work on my current installation how it will work ?


2. Also, does it mean that I will have to restart the whole workspace (if I retrieve the value from environment variable from inside the container) while the plugins can be dynamically loaded at runtime without a restart ?

**Comment by benoitf** [comment]
it's maybe implementation details for you but let say I work on implementing the calls for theia plugins. And for now it's not elegant at all for clients. (including the work I need to do)

> .how can I see the description of these two theia plugins ?
What for? Everything that user need to see or interact with is in ChePlagin metadata or in Feature description.

1. Where I specify in my .car file that I have two theia plugins ? I suppose dropping two zip files is not enough. Is it in the plugin-metadata.xml that it is ?

2. if you're saying that only feature metadata is enough, let say you pickup two random zip files and you add in into the .car file.
Client need to know which is the engine version supported for these two theia plugins. so somehow, registry has parsed it and can deliver that information to a client. (and not only an URL)


3.
 >> Also, does it mean that I will have to restart the whole workspace (if I retrieve the value from environment variable from inside the container) while the plugins can be dynamically loaded at runtime without a restart ?

> yes.

why the hell I would have a restart. We're talking about a model and even the model already is forcing a restart that could be easily avoided. A restart should not be part of the model.

**Comment by benoitf** [comment]
Here is an example of vscode marketplace
if I take a random page like
https://marketplace.visualstudio.com/items?itemName=DotJoshJohnson.xml

I have a link https://marketplace.visualstudio.com/_apis/public/gallery/publishers/DotJoshJohnson/vsextensions/xml/1.9.2/vspackage

where I can grab the plugin zip directly.

**Comment by skabashnyuk** [comment]
>Where I specify in my .car file that I have two theia plugins ? I suppose dropping two zip files is not enough. Is it in the plugin-metadata.xml that it is ?
>if you're saying that only feature metadata is enough, let say you pickup two random zip files and you add in into the .car file.
>Client need to know which is the engine version supported for these two theia plugins. so somehow, registry has parsed it and can deliver that information to a client. (and not only an URL)


In our case, all that we can deliver is a parameter for container. More concrate environment variables. In our case, if you need to deliver several artifacts to some container it can be made in this way.
```yaml
apiVersion: v1
kind: CheFeature
metadata:
 name: che-theia-github
spec:
 version: 0.0.1
 description: Theia's Github plugin
 services:
   - name : org.eclipse.che.theia-ide
     version : 0.0.1
     parameters:
         - name: THEIA_PLUGINS
           value: ${url:che-theia-github.tar.gz}
         - name: GITHUB_CLI_TOOL_KEY
           value: ${url:another_importan_file.tar.gz}   
```

>why the hell I would have a restart. We're talking about a model and even the model already is forcing a restart that could be easily avoided. A restart should not be part of the model.


I see that this question is quite important for you and I'm happy to participate with it. In case if you can provide some more details how you see it will going to happen. Some links to specification would be welcome. AFAIK what I know from past decisions that were confirmed by @gorkem  we **ARE NOT GOING TO SUPPORT** enabling/disabling plugins in case if a workspace is running at least in phase 1. Since in general case it requires starting of the additional containers in the same pod. If I've missed something please let me know.

**Comment by skabashnyuk** [comment]
>I have a link https://marketplace.visualstudio.com/_apis/public/gallery/publishers/DotJoshJohnson/vsextensions/xml/1.9.2/vspackage

>where I can grab the plugin zip directly.

Ok we can imagine ```GET``` method to get ```.cfa```  file back. Would it work for you?

**Comment by benoitf** [comment]
ok let's move out lifecycle for now for plugins

I'm more interested in having a GET method for each 'theia plugin'
it could work with the cfa but I presume that if archive is containing screenshots, etc it may not be interesting at all

GET `plugin/version/<something-to-get-what is inside the cfa like theia plugins>`

GET /plugin/version/theia I have array of theia plugins for this che plugin (plugin1 and plugin2)
GET /plugin/version/theia/plugin1/ --> I have metadata of the theiaPlugin 1
GET /plugin/version/theia/plugin1/package I have the zip file of theiaPlugin1
GET /plugin/version/theia/plugin2/ --> I have metadata of the theiaPlugin 2
GET /plugin/version/theia/plugin2/package --> I have zip of the theiaPlugin 2

(what is metadata of plugin1, it's the package.json file including the engine required to run, if it's a frontend or backend plugins, etc)

**Comment by benoitf** [comment]
let's refresh as well:
-plugins, market place and workspace.next are three working items. I should be able to work with marketplace without having workspace.next complete and workspace.next can still work without plugins.

--> I still need to get engine level of a theia plugin when I want to enable or disable a feature on top of my workspace or to ensure that it will work when publishing it. (two checks)

--> right now there is no workspace.next running so I will still require to get access to the theia plugins 
You can hide in features, services or anything but binary should be available and metadata of theia plugin should be available as well.

**Comment by skabashnyuk** [comment]
> -->plugins, market place and workspace.next are three working items. I should be able to work with marketplace without having workspace.next complete and workspace.next can still work without plugins.
> --> I still need to get engine level of a theia plugin when I want to enable or disable a feature on top of my workspace or to ensure that it will work when publishing it. (two checks)
> --> right now there is no workspace.next running so I will still require to get access to the theia plugins

Not sure all of that are possible. What I can see that we all should do our best to move forward plugins, market place and workspace.next in parallel
and not to make problems to each other.

> You can hide in features, services or anything but binary should be available and metadata of theia plugin should be available as well.

Ok. Tell me what and how are you going to do with them? How its related to eclipse/che-theia:nightly, our image that is going to run Theia inside of Eclipse Che

**Comment by benoitf** [comment]
>>You can hide in features, services or anything but binary should be available and metadata of theia >>plugin should be available as well.

>Ok. Tell me what and how are you going to do with them? How its related to eclipse/che-theia:nightly, >our image that is going to run Theia inside of Eclipse Che

image could grab the plugins and install them or not install them due to an incompatible engine version (by getting theia plugin's metadata through registry)
Also theia-nightly image could show (or in addition to dashboard) could display all compatible features for theia

**Comment by skabashnyuk** [comment]
>image could grab the plugins and install them or not install them due to an incompatible engine version (by getting theia plugin's metadata through registry)

So I guess, this will work for you. When we provide a comma-separated list of plugins as an environment variable of theia container. BTW in future, we can optimize this process and convert this links to paths to some attached volume where we can put all or some most frequently used artifacts.
```
     parameters:
         - name: THEIA_PLUGINS
           value: http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github.tar.gz,http://localhost:3000/files/org.eclipse.che.che-theia-github/che-theia-github2.tar.gz
```
The only one missing part of what I understand is the server who can serve artifacts over http. @flodernt am I correct?


>Also theia-nightly image could show (or in addition to dashboard) display all compliant features for theia by asking the registry

We can but what the purpose of functionality duplication?
From @stevans spec: "the plugin listed under workspace's settings." That means that we will have someplace out of Theia IDE, for instance on the dashboard, where a user can select
Plugin/Feature.
Second point: even if we display them, not all of them will be possible to enable without restarting. To be more concrete all serious plugins like Language support, some tooling that
needs external dependencies to be installed, etc.

**Comment by benoitf** [comment]
I think we're stuck on the discussion so I think we should move to a call after summit.

What I mean is that registry should expose other data, so one view is for what you want to implement, but there are still other views/data access to the registry. I could ask and get more stuff from the registry API on metadata (including theia plugin's metadata)

>> Also theia-nightly image could show (or in addition to dashboard) display all compliant features for theia by asking the registry
>We can but what the purpose of functionality duplication?
From @slemeur spec: "the plugin listed under workspace's settings." That means that we will have someplace out of Theia IDE, for instance on the dashboard, where a user can select
Plugin/Feature.
For example dashboard will probably need to ask as well what are the theia plugins inside the features to avoid some collision or to have a full dependency graph in order to say that it won't work at runtime if we choose these plugins for example.
So there is no duplication there.


>Second point: even if we display them, not all of them will be possible to enable without restarting. To be more concrete all serious plugins like Language support, some tooling that
needs external dependencies to be installed, etc.

yes but it should not be the model that imply the choice. Also if my che plugin only contains a theia plugin it is possible to install it without restarting the workspace and UX is a lot better.

**Comment by dneary** [comment]
Has any thought been put into the server-side platform for Eclipse Che add-ons? Is the plan to use Eclipse Foundation infrastructure? If the goal is to have something specifically for Che, has anyone had a look at the [Mozilla Add-ons server](http://addons-server.readthedocs.io/en/latest/index.html) yet? http://addons-server.readthedocs.io/en/latest/index.html It is open source, container deployed, has a separation of front-end and back-end, and has a lot of features that are very desirable for an add-ons platform (ratings, comments, portal to show the most popular add-ons, etc).

**Comment by slemeur** [comment]
@dneary : Thanks for sharing. This would cover what we need and even more. We should definitely consider using it!

**Comment by vparfonov** [comment]
@dneary Thanks for sharing this information https://github.com/eclipse/che/issues/9863

**Comment by sleshchenko** [comment]
@slemeur Looks like no any activities in this issue for a long time, but as far as I know, there are few issues for Plugin Registry UI. Should this epic issue be closed or updated to contains actual plans around it?

**Comment by slemeur** [comment]
Yes we can close this issue.

Most of the work has been handle with the Plugin Registry
https://github.com/eclipse/che-plugin-registry
